pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentTarPushJob.java

```diff
@@ -47,9 +54,57 @@ protected boolean isDataFile(String fileName) {
   public void run()
       throws Exception {
     FileSystem fileSystem = FileSystem.get(_conf);
+    List<Path> segmentsToPush = getDataFilePaths(_segmentPattern);
     try (ControllerRestApi controllerRestApi = getControllerRestApi()) {
-      controllerRestApi.pushSegments(fileSystem, getDataFilePaths(_segmentPattern));
+      // TODO: Deal with invalid prefixes in the future
+
+      List<String> allSegments = controllerRestApi.getAllSegments("OFFLINE");
+
+      controllerRestApi.pushSegments(fileSystem, segmentsToPush);
+
+      if (_deleteExtraSegments) {
+        controllerRestApi.deleteSegmentUris(getSegmentsToDelete(allSegments, segmentsToPush));
+      }
+    }
+  }
+
+  /**
+   * Deletes extra segments after pushing to the controller
+   * @param allSegments all segments on the controller for the table
+   * @param segmentsToPush segments that will be pushed to the controller
+   * @throws IOException
+   */
+  public List<String> getSegmentsToDelete(List<String> allSegments, List<Path> segmentsToPush) {
+    Set<String> uniqueSegmentPrefixes = new HashSet<>();
+
+    // Get all relevant segment prefixes that we are planning on pushing
+    List<String> segmentsToPushNames = segmentsToPush.stream().map(s -> s.getName()).collect(Collectors.toList());
+    for (String segmentName : segmentsToPushNames) {
+      String segmentNamePrefix = removeSequenceId(segmentName);
+      uniqueSegmentPrefixes.add(segmentNamePrefix);
+    }
+
+    List<String> relevantSegments = new ArrayList<>();
+    // Get relevant segments already pushed that we are planning on refreshing
+    for (String segmentName : allSegments) {
+      if (uniqueSegmentPrefixes.contains(removeSequenceId(segmentName))) {
+        relevantSegments.add(segmentName);
+      }
     }
+
+    relevantSegments.removeAll(segmentsToPushNames);
```

Comment:
This will be a O(N^2) operation and potentially more heap heavy because of constant over-writing.
I think we should use a hashset (or any alternative) to make this constant time operation